File: ANSv4/Catalogo3/classes/EnvioRendicion.vb
Option Explicit On

Public Class EnvioRendicion

    ' Define como se llama este modulo para el control de errores
    Private Const m_sMODULENAME_ As String = "clsEnvioRendicion"

    ' Esta clase realiza una consulta al servidor web y obtiene
    ' el numero de ip del servidor privado.
    Private Cliente As WSRendicion.RendicionSoapClient
    Private WebServiceInicializado As Boolean
    Private m_MacAddress As String
    Private m_ip As String

    Private m_NroRendicion As String
    Private m_IdViajante As String
    Private m_F_Rendicion As String
    Private m_Observaciones As String
    Private m_Efectivo As Double
    Private m_Dolares As Single
    Private m_Euros As Single
    Private m_ChequesCant As Byte
    Private m_ChequesMonto As Double
    Private m_CertificadosCant As Byte
    Private m_CertificadosMonto As Double

    Private m_DetalleValores As String
    Private m_DetalleRecibos As String

    Private DatosObtenidos As Boolean


    'Public Property Get Inicializado() As Boolean
    'Inicializado = WebServiceInicializado
    '    End Property

    '    Public Function ObtenerDatos(ByVal NroRendicion As String)

    '        DatosObtenidos = False

    '        Dim rs_Ren As ADODB.Recordset
    '        Dim rs_RenValores As ADODB.Recordset
    '        Dim rs_RenRecibos As ADODB.Recordset

    '        rs_Ren = New ADODB.Recordset
    '        rs_RenValores = New ADODB.Recordset
    '        rs_RenRecibos = New ADODB.Recordset

    '        rs_Ren = adoModulo.adoComando(vg.Conexion, "SELECT * FROM v_Rendicion WHERE Nro='" & NroRendicion & "'")
    '        rs_RenValores = adoModulo.adoComando(vg.Conexion, "SELECT * FROM  v_RendicionValores1 WHERE Nro='" & NroRendicion & "'")
    '        rs_RenRecibos = adoModulo.adoComando(vg.Conexion, "EXECUTE v_Rendicion_Recibos_rpt '" & Right(NroRendicion, 8) & "'")

    '        m_NroRendicion = NroRendicion
    '        m_IdViajante = Format(Trim(rs_Ren("IDCliente")), "000000")
    '        m_F_Rendicion = Mid(rs_Ren("F_Rendicion"), 7, 4) & Mid(rs_Ren("F_Rendicion"), 4, 2) & Mid(rs_Ren("F_Rendicion"), 1, 2)
    '        m_Observaciones = Trim(rs_Ren("Descripcion"))

    '        m_Efectivo = Format(Trim(rs_Ren("Efectivo_Monto") * 100), "00000000000000000")
    '        m_Dolares = Format(Trim(rs_Ren("Dolar_Cantidad") * 100), "00000000000000000")
    '        m_Euros = Format(Trim(rs_Ren("Euros_Cantidad") * 100), "00000000000000000")
    '        m_ChequesMonto = Format(Trim(rs_Ren("Cheques_Monto") * 100), "00000000000000000")
    '        m_CertificadosMonto = Format(Trim(rs_Ren("Certificados_Monto") * 100), "00000000000000000")
    '        m_ChequesCant = Format(rs_Ren("Cheques_Cantidad"), "00")
    '        m_CertificadosCant = Format(rs_Ren("Certificados_Cantidad"), "00")

    '        'rs_RenValores.MoveFirst
    '        m_DetalleValores = ""
    '        Do Until rs_RenValores.EOF
    '0:          m_DetalleValores = m_DetalleValores & rs_RenValores("Bco_Dep_Tipo") & ","
    '1:          m_DetalleValores = m_DetalleValores & Mid(rs_RenValores("Bco_Dep_Fecha"), 7, 4) & Mid(rs_RenValores("Bco_Dep_Fecha"), 4, 2) & Mid(rs_RenValores("Bco_Dep_Fecha"), 1, 2) & ","
    '2:          m_DetalleValores = m_DetalleValores & padLR(rs_RenValores("Bco_Dep_Numero"), 10, "0", 1) & ","
    '3:          m_DetalleValores = m_DetalleValores & Format(Trim(rs_RenValores("Bco_Dep_Monto") * 100), "00000000000000000") & ","
    '4:          m_DetalleValores = m_DetalleValores & Format(rs_RenValores("Bco_Dep_Ch_Cantidad"), "00") & ","
    '5:          m_DetalleValores = m_DetalleValores & padLR(Trim(" " & rs_RenValores("N_Cheque")), 15, " ", 2) & ","
    '6:          m_DetalleValores = m_DetalleValores & Format(rs_RenValores("IdBanco"), "000") & ";"
    '            rs_RenValores.MoveNext()
    '        Loop

    '        'rs_RenRecibos.MoveFirst
    '        m_DetalleRecibos = ""
    '        Do Until rs_RenRecibos.EOF
    '            '0:
    '            m_DetalleRecibos = m_DetalleRecibos & rs_RenRecibos!nroRecibo & ","
    '            '1:
    '            m_DetalleRecibos = m_DetalleRecibos & Mid(rs_RenRecibos("F_Recibo"), 7, 4) & Mid(rs_RenRecibos("F_Recibo"), 4, 2) & Mid(rs_RenRecibos("F_Recibo"), 1, 2) & ","
    '            '2:
    '            m_DetalleRecibos = m_DetalleRecibos & Format(Trim(rs_RenRecibos!Nro_Cuenta), "000000") & ","
    '            '3:
    '            m_DetalleRecibos = m_DetalleRecibos & Format(Trim(rs_RenRecibos!Efectivo * 100), "00000000000000000") & ","
    '            '4:
    '            m_DetalleRecibos = m_DetalleRecibos & Format(Trim(rs_RenRecibos!Divisas_Dolares * 100), "00000000000000000") & ","
    '            '5:
    '            m_DetalleRecibos = m_DetalleRecibos & Format(Trim(rs_RenRecibos!Divisas_Euros * 100), "00000000000000000") & ","
    '            '6:
    '            m_DetalleRecibos = m_DetalleRecibos & Format(Trim(rs_RenRecibos!Cheques_Total * 100), "00000000000000000") & ","
    '            '7:
    '            m_DetalleRecibos = m_DetalleRecibos & Format(rs_RenRecibos!Cheques_Cantidad, "00") & ","
    '            '8:
    '            m_DetalleRecibos = m_DetalleRecibos & Format(Trim(rs_RenRecibos!Certificados_Total * 100), "00000000000000000") & ","
    '            '9:
    '            m_DetalleRecibos = m_DetalleRecibos & Format(rs_RenRecibos!Certificados_Cantidad, "00") & ","
    '            '10:
    '            m_DetalleRecibos = m_DetalleRecibos & Format(Trim(rs_RenRecibos!TotalRecibo * 100), "00000000000000000") & ";"

    '            rs_RenRecibos.MoveNext()
    '        Loop

    '        rs_Ren = Nothing
    '        rs_RenValores = Nothing
    '        rs_RenRecibos = Nothing

    '        DatosObtenidos = True

    '    End Function

    Public Function EnviarRendicion() As Long

        On Error GoTo ErrorHandler

        Dim Cancel As Boolean

        Cancel = False

        Dim resultado As Long

        If Not WebServiceInicializado Then
            Cancel = True
        End If

        If Not Cancel Then

            If vg.TranActiva Is Nothing Then
                vg.TranActiva = vg.Conexion.BeginTransaction
            End If

            resultado = Cliente.EnviarRendicion(m_MacAddress, m_NroRendicion, m_IdViajante, m_F_Rendicion, m_Observaciones, _
                        m_Efectivo, m_Dolares, m_Euros, m_ChequesCant, m_ChequesMonto, m_CertificadosCant, m_CertificadosMonto, _
                        m_DetalleValores, m_DetalleRecibos)

            If resultado = 0 Then
                '                adoModulo.adoComandoIU(vg.Conexion, "EXEC usp_Rendicion_Transmicion_Upd '" & Right(m_NroRendicion, 8) & "'")
                If Not vg.TranActiva Is Nothing Then
                    vg.TranActiva.Commit()
                    vg.TranActiva = Nothing
                End If
            Else
                If Not vg.TranActiva Is Nothing Then
                    vg.TranActiva.Rollback()
                    vg.TranActiva = Nothing
                End If
            End If

            EnviarRendicion = resultado
        Else
            EnviarRendicion = -1
        End If

        Exit Function

ErrorHandler:
        Err.Raise(Err.Number, Err.Source, Err.Description)

    End Function

    Public Sub Inicializar(ByVal ipAddress As String, _
                           ByVal ipAddressIntranet As String, _
                           ByVal MacAddress As String)

        Dim Conectado As Boolean
        Conectado = My.Computer.Network.Ping(ipAddress, 5000)
        If Not Conectado Then
            Conectado = My.Computer.Network.Ping(ipAddressIntranet, 5000)
        End If

        On Error GoTo ErrorHandler

        If Conectado Then
            If Not WebServiceInicializado Then
                Cliente = New WSRendicion.RendicionSoapClient("", "http://" & ipAddress & "/wsCatalogo3/Rendicion.asmx?wsdl")
                If Trim(Funciones.modINIs.ReadINI("datos", "proxy")) = "1" Then
                    '                    Cliente.ConnectorProperty("ProxyServer") = Funciones.modINIs.ReadINI("datos", "proxyserver")
                End If

                m_MacAddress = MacAddress
                m_ip = ipAddress
                WebServiceInicializado = True
            End If
        Else
            WebServiceInicializado = False
        End If

        Exit Sub

ErrorHandler:

        If Err.Number = -2147024809 Then
            ' Intento con el ip interno
            Cliente = New WSRendicion.RendicionSoapClient("", "http://" & ipAddressIntranet & "/wsCatalogo3/Rendicion.asmx?wsdl")
            If Trim(Funciones.modINIs.ReadINI("datos", "proxy")) = "1" Then
                '                Cliente.ConnectorProperty("ProxyServer") = Funciones.modINIs.ReadINI("datos", "proxyserver")
            End If

            m_MacAddress = MacAddress
            m_ip = ipAddressIntranet
            WebServiceInicializado = True
            Err.Clear()
        Else
            Err.Raise(Err.Number, Err.Source, Err.Description)
        End If

    End Sub


End Class

File: ANSv4/Catalogo4/_rendiciones/EnvioRendicion.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
//using Microsoft.VisualBasic;

namespace Catalogo._rendicion
{
    public class EnvioRendicion
    {
        // Esta clase realiza una consulta al servidor web y obtiene
        // el numero de ip del servidor privado.
        private RendicionWS.Rendicion Cliente;
        private bool WebServiceInicializado;
        private string m_MacAddress;

        private string m_ip;
        private string m_NroRendicion;
        private string m_IdViajante;
        private string m_F_Rendicion;
        private string m_Observaciones;
        private float m_Efectivo;
        private float m_Dolares;
        private float m_Euros;
        private byte m_ChequesCant;
        private float m_ChequesMonto;
        private byte m_CertificadosCant;

        private float m_CertificadosMonto;
        private string m_DetalleValores;

        private string m_DetalleRecibos;

        private bool DatosObtenidos;

        private System.Data.OleDb.OleDbConnection Conexion1;

        public EnvioRendicion(System.Data.OleDb.OleDbConnection conexcion, string ipAddress, string ipAddressIntranet, string MacAddress, bool usaProxy, string proxyServerAddress)
        {
            Inicializar(ipAddress, ipAddressIntranet, MacAddress, usaProxy, proxyServerAddress);
            Conexion1 = conexcion;
        }

        public bool Inicializado
        {
            get { return WebServiceInicializado; }
        }


        public void ObtenerDatos(string NroRendicion)
        {
            DatosObtenidos = false;

            System.Data.OleDb.OleDbDataReader Ren = null;
            System.Data.OleDb.OleDbDataReader RenValores = null;
            System.Data.OleDb.OleDbDataReader RenRecibos = null;

            Ren = Funciones.oleDbFunciones.Comando(Conexion1, "SELECT * FROM v_Rendicion WHERE Nro='" + NroRendicion + "'");
            RenValores = Funciones.oleDbFunciones.Comando(Conexion1, "SELECT * FROM  v_RendicionValores1 WHERE Nro='" + NroRendicion + "'");
            RenRecibos = Funciones.oleDbFunciones.Comando(Conexion1, "EXECUTE v_Rendicion_Recibos_rpt '" + NroRendicion.Substring(NroRendicion.Length - 8) + "'");

            m_NroRendicion = NroRendicion;
            Ren.Read();
            m_IdViajante = Ren["IDCliente"].ToString().Trim().PadLeft(6, '0');
            m_F_Rendicion = Ren["F_Rendicion"].ToString().Substring(7, 4) +
                Ren["F_Rendicion"].ToString().Substring(4, 2) +
                Ren["F_Rendicion"].ToString().Substring(1, 2);
            m_Observaciones = Ren["Descripcion"].ToString().Trim();

            m_Efectivo = float.Parse(Ren["Efectivo_Monto"].ToString()) * 100;
            m_Dolares = float.Parse(Ren["Dolar_Cantidad"].ToString()) * 100;
            m_Euros = float.Parse(Ren["Euros_Cantidad"].ToString()) * 100;
            m_ChequesMonto = float.Parse(Ren["Cheques_Monto"].ToString()) * 100;
            m_CertificadosMonto = float.Parse(Ren["Certificados_Monto"].ToString()) * 100;
            m_ChequesCant = byte.Parse(Ren["Cheques_Cantidad"].ToString());
            m_CertificadosCant = byte.Parse(Ren["Certificados_Cantidad"].ToString());

            if (RenValores.HasRows)
            {
                m_DetalleValores = "";
                while (RenValores.Read())
                {
                    m_DetalleValores += RenValores["Bco_Dep_Tipo"].ToString() + ",";
                    m_DetalleValores += RenValores["Bco_Dep_Fecha"].ToString().Substring(7, 4)
                        + RenValores["Bco_Dep_Fecha"].ToString().Substring(4, 2)
                        + RenValores["Bco_Dep_Fecha"].ToString().Substring(1, 2) + ",";
                    m_DetalleValores += RenValores["Bco_Dep_Numero"].ToString().Substring(10) + ",";
                    m_DetalleValores += (float.Parse(RenValores["Bco_Dep_Monto"].ToString()) * 100).ToString().Trim().PadLeft(17, '0') + ",";
                    m_DetalleValores += RenValores["Bco_Dep_Ch_Cantidad"].ToString().Trim().PadLeft(2, '0') + ",";
                    m_DetalleValores += RenValores["N_Cheque"].ToString().Trim().PadLeft(15, ' ') + ",";
                    m_DetalleValores += RenValores["IdBanco"].ToString().Trim().PadLeft(3, '0') + ";";
                }
            }

            if (RenRecibos.HasRows)
            {
                m_DetalleRecibos = "";
                while (RenRecibos.Read())
                {
                    m_DetalleRecibos += RenRecibos["nroRecibo"].ToString() + ",";
                    m_DetalleRecibos += RenRecibos["F_Recibo"].ToString().Substring(7, 4)
                        + RenRecibos["F_Recibo"].ToString().Substring(4, 2)
                        + RenRecibos["F_Recibo"].ToString().Substring(1, 2) + ",";
                    m_DetalleRecibos += RenRecibos["Nro_Cuenta"].ToString().Trim().PadLeft(6, '0') + ",";
                    m_DetalleRecibos += (float.Parse(RenRecibos["Efectivo"].ToString()) * 100).ToString().Trim().PadLeft(17, '0') + ",";
                    m_DetalleRecibos += (float.Parse(RenRecibos["Divisas_Dolares"].ToString()) * 100).ToString().Trim().PadLeft(17, '0') + ",";
                    m_DetalleRecibos += (float.Parse(RenRecibos["Divisas_Euros"].ToString()) * 100).ToString().Trim().PadLeft(17, '0') + ",";
                    m_DetalleRecibos += (float.Parse(RenRecibos["Cheques_Total"].ToString()) * 100).ToString().Trim().PadLeft(17, '0') + ",";
                    m_DetalleRecibos += RenRecibos["Cheques_Cantidad"].ToString().Trim().PadLeft(2, '0') + ",";
                    m_DetalleRecibos += (float.Parse(RenRecibos["Certificados_Total"].ToString()) * 100).ToString().Trim().PadLeft(17, '0') + ",";
                    m_DetalleRecibos += RenRecibos["Certificados_Cantidad"].ToString().Trim().PadLeft(2, '0') + ",";
                    m_DetalleRecibos += (float.Parse(RenRecibos["TotalRecibo"].ToString()) * 100).ToString().Trim().PadLeft(17, '0') + ";";
                }
            }

            Ren = null;
            RenValores = null;
            RenRecibos = null;

            DatosObtenidos = true;

        }

        public long EnviarRendicion()
        {
            bool Cancel = false;
            long resultado = 0;

            if (!WebServiceInicializado)
            {
                Cancel = true;
            }


            if (!Cancel)
            {
                if (Global01.TranActiva == null)
                {
                    Global01.TranActiva = Conexion1.BeginTransaction();
                }

                resultado = Cliente.EnviarRendicion(m_MacAddress, m_NroRendicion, m_IdViajante, m_F_Rendicion, m_Observaciones,
                    m_Efectivo.ToString(), m_Dolares.ToString(), m_Euros.ToString(),
                    m_ChequesCant.ToString(), m_ChequesMonto.ToString(),
                m_CertificadosCant.ToString(), m_CertificadosMonto.ToString(), m_DetalleValores, m_DetalleRecibos);

                if (resultado == 0)
                {
                    Funciones.oleDbFunciones.ComandoIU(Conexion1, "EXEC usp_Rendicion_Transmicion_Upd '" + m_NroRendicion.Substring(m_NroRendicion.Length - 8) + "'");
                    if (Global01.TranActiva != null)
                    {
                        Global01.TranActiva.Commit();
                        Global01.TranActiva = null;
                    }
                }
                else
                {
                    if (Global01.TranActiva != null)
                    {
                        Global01.TranActiva.Rollback();
                        Global01.TranActiva = null;
                    }
                }

                return resultado;
            }
            else
            {
                return -1;
            }
        }


        public void Inicializar(string ipAddress, string ipAddressIntranet, string MacAddress, bool usaProxy, string proxyServerAddress)
        {
            bool Conectado = false;
            Conectado = util.SimplePing.ping(ipAddress, 5000);
            if (!Conectado)
            {
                Conectado = util.SimplePing.ping(ipAddressIntranet, 5000);
            }

            try
            {

                if (Conectado)
                {
                    if (!WebServiceInicializado)
                    {
                        Cliente = new RendicionWS.Rendicion();
                        Cliente.Url = "http://" + ipAddress + "/wsCatalogo4/Rendicion.asmx?wsdl";
                        if (usaProxy)
                        {
                            Cliente.Proxy = new System.Net.WebProxy(proxyServerAddress);
                        }

                        m_MacAddress = MacAddress;
                        m_ip = ipAddress;
                        WebServiceInicializado = true;
                    }
                }
                else
                {
                    WebServiceInicializado = false;
                }
            }
            catch (Exception ex)
            {
                if (System.Runtime.InteropServices.Marshal.GetExceptionCode() == -2147024809)
                {
                    //errhandler:
                    //        If Err.Number = -2147024809 Then
                    // Intento con el ip interno
                    Cliente = new RendicionWS.Rendicion();
                    Cliente.Url = "http://" + ipAddressIntranet + "/wsCatalogo4/Rendicion.asmx?wsdl";
                    if (usaProxy)
                    {
                        Cliente.Proxy = new System.Net.WebProxy(proxyServerAddress);
                    }

                    m_MacAddress = MacAddress;
                    m_ip = ipAddressIntranet;
                    WebServiceInicializado = true;
                }
                else
                {
                    throw ex;
                }
            }
        }
    }
}